app/models/manageiq/providers/amazon/cloud_manager/event_parser.rb

```diff
@@ -30,33 +59,132 @@ def self.parse_cloud_watch_alarm_event!(event, event_hash)
     event_hash[:availability_zone_ems_ref] = nil # Can't get it
   end
 
-  def self.event_to_hash(event, ems_id)
-    event_hash = {
-      :event_type => event["eventType"],
-      :source     => "AMAZON",
-      :full_data  => event,
-      :ems_id     => ems_id
-    }
+  def self.parse_vm_ref(event)
+    resource_type = event["configurationItem"]["resourceType"]
+    # other ways to find the VM?
+    event.fetch_path("configurationItem", "resourceId") if resource_type == "AWS::EC2::Instance"
+  end
 
-    unless %i(config cloud_watch_api cloud_watch_ec2 cloud_watch_alarm).include?(event["event_source"])
-      raise "Unsupported event source #{event["event_source"]}"
+  def self.parse_availability_zone_ref(event)
+    event.fetch_path("configurationItem", "availabilityZone")
+  end
+
+  def self.parse_ems_event_targets(event)
+    available_targets = self.available_targets || {}
+
+    case event.full_data["event_source"]
+    when :cloud_watch_api
+      collect_cloudwatch_api_references!(available_targets, event.full_data.fetch_path("detail", "requestParameters") || {})
```

Comment:
all those bang methods. It's hard to tell what they actually modify.
This screams for extraction of an object 🛠